pkg/smart/common.go

```diff
@@ -62,14 +59,39 @@ func detectSCSIType(name string) (Dev, error) {
 	return &device, nil
 }
 
-// readDeviceCapacity sends a SCSI READ CAPACITY(10) command to a device and
-// returns the capacity in bytes. If this command fails due to a capacity
-// overflow, it retries a SCSI READ CAPACITY(16) command.
+// isReadDeviceCapacity16Supported checks if the SCSI READ CAPACITY(16)
+// command is supported by the device.
+func (d *SCSIDev) isReadDeviceCapacity16Supported() (bool, error) {
+	respBuf := make([]byte, 20)
+
+	cdb := CDB12{}
+	cdb[0] = SCSIReportSupportedOpCodes
+	cdb[1] = SCSIReportSupportedOpCodesSA
+	cdb[2] = 0x02
+	cdb[3] = SCSIReadCapacity16
+	binary.BigEndian.PutUint16(cdb[4:], uint16(SCSIReadCapacitySA))
+	binary.BigEndian.PutUint32(cdb[6:], uint32(len(respBuf)))
+
+	if err := d.sendSCSICDB(cdb[:], &respBuf); err != nil {
+		return false, err
+	}
+
+	switch respBuf[1] & 0x07 {
+	case 0x03, 0x05:
+		return true, nil
+	default:
+		return false, nil
+	}
+}
+
+// readDeviceCapacity returns the devices capacity in bytes.
 func (d *SCSIDev) readDeviceCapacity() (uint64, error) {
-	if DeviceCapacity, err := d.readDeviceCapacity10(); err == errReadCapacityOverflow {
+	if supported16, err := d.isReadDeviceCapacity16Supported(); err != nil {
```

Comment:
on `err`, would it be good to do readDeviceCapacity10()?

Reply:
I don't know in which scenario `isReadDeviceCapacity16Supported()` might fail, but `readDeviceCapacity10()` still work. But if it does, I assume `readDeviceCapacity10()` would need to check for the capacity overflow again.